File: ios/CustomComponent/RNTMapManager.m
//
//  RNTMapManager.m
//  CustomComponent
//
//  Created by Poorva Singh on 08/08/18.
//  Copyright © 2018 Facebook. All rights reserved.
//

#import "React/RCTViewManager.h"
#import <MapKit/MapKit.h>
#import "RCTConvert+MapKit.h"

@interface RCT_EXTERN_MODULE(RNTMapManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(zoomEnabled, BOOL)
RCT_CUSTOM_VIEW_PROPERTY(region, MKCoordinateRegion, MKMapView) {
  [view setRegion:json ? [RCTConvert MKCoordinateRegion:json]: defaultView.region animated:YES];
}

RCT_EXPORT_VIEW_PROPERTY(onRegionChange, RCTBubblingEventBlock)

@end

File: ios/CustomComponent/RNTMapManager.swift
//
//  RNTMapManager.swift
//  CustomComponent
//
//  Created by Poorva Singh on 08/08/18.
//  Copyright © 2018 Facebook. All rights reserved.
//

import MapKit

@objc(RNTMapManager)
class RNTMapManager: RCTViewManager {
  
  override func view() -> UIView {
    let map = RNTMapView()
    map.delegate = self
    return map
  }
}

extension RNTMapManager: MKMapViewDelegate {
 
  func mapView(_ mapView: MKMapView, regionDidChangeAnimated animated: Bool) {
    guard let mapView = mapView as? RNTMapView, let regionChange = mapView.onRegionChange else {
      return
    }
    
    let region = mapView.region as MKCoordinateRegion
    regionChange( ["region" : [
      "latitude": region.center.latitude,
      "longitude": region.center.longitude,
      "latitudeDelta": region.span.latitudeDelta,
      "longitudeDelta": region.span.longitudeDelta
      ]
    ]
    )
  }
}
